test/service_test.cc

```diff
@@ -0,0 +1,458 @@
+#include <chrono>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <sys/time.h>
+#include <thread>
+#include <vector>
+
+#include <glog/logging.h>
+#include "gtest/gtest.h"
+
+#include "service_client_lib.h"
+#include "service_data_structure.h"
+
+namespace {
+
+uint64_t PrintId(const std::string &id) {
+  return *(reinterpret_cast<const uint64_t*>(id.c_str()));
+}
+
+const size_t kNumOfUsersPreset = 10;
+const size_t kNumOfUsersTotal = 20;
+const char *kShortText = "short";
+const char *kLongText = "longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong";
+
+// This test cases on the `ServiceDataStructure` to check whether their
+// interfaces work correctly.
+class ServiceTestDataStructure : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    chirp_connect_backend::backend_client_.reset(new BackendClientDebug());
+
+    // Set up users to be used in the sub-tests
+    for(size_t i = 0; i < kNumOfUsersTotal; ++i) {
+      user_list_.push_back(std::string("user") + std::to_string(i));
+      if (i < kNumOfUsersPreset) {
+        service_data_structure_.UserRegister(user_list_[i]);
+      }
+    }
+  }
+
+  // The container to store the usernames
+  std::vector<std::string> user_list_;
+
+  // The object that this test is going to test
+  ServiceDataStructure service_data_structure_;
+};
+
+// This tests on the `UserRegister` and `Login` in `ServiceDataStructure`
+TEST_F(ServiceTestDataStructure, UserRegisterAndLoginTest) {
+  // Try to register existed usernames which already done in the `SetUp()`
+  for(size_t i = 0; i < kNumOfUsersPreset; ++i) {
+    // ServiceDataStructure::ReturnCodes
+    auto ret = service_data_structure_.UserRegister(user_list_[i]);
+    // This should fail since the username specified has already been registered in the `SetUp` process above
+    EXPECT_EQ(ServiceDataStructure::USER_EXISTS, ret);
```

Comment:
Does it make sense to run this code in a loop? it seems redundant to execute the same scenario 10 times

Reply:
The 10 is just a random number I picked. Just wanted to check if this system can accommodate some number of users.